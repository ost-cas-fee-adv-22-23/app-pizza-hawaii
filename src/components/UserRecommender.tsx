import React, { FC, useEffect, useState } from 'react';
import { Grid, Headline } from '@smartive-education/pizza-hawaii';
import { services } from '../services';
import { useSession } from 'next-auth/react';
import { UserCard } from './UserCard';
import { TUser } from '../types';

interface TUserRecommender {
	currentUserId: string;
}

export const UserRecommender: FC<TUserRecommender> = (props) => {
	const [isLoading, setIsLoading] = useState(true);
	const [recommendedUsers, setRecommendedUsers] = useState([] as TUser[]);
	const { data: session } = useSession();
	const { currentUserId } = props;
	const accessToken = session?.accessToken;

	// optional TODO: should we fetch recommended users via API ?
	// TODO do a fetch with /api/users/getall

	useEffect(() => {
		if (accessToken) {
			const fetchRecommendedUsers = async (accessToken: string) => {
				const recommendedUsers = await services.users.getUsers({
					limit: 7,
					offset: 0,
					accessToken,
				});
				setRecommendedUsers(recommendedUsers.users);
			};
			fetchRecommendedUsers(accessToken);
			setIsLoading(false);
		}
	}, [accessToken]);

	// exclude current user from recommended users
	const pureRecommendedUsers = recommendedUsers.filter((user) => user.id !== currentUserId);

	return (
		<>
			<Headline as="h2" level={3}>
				Empfohlene User
			</Headline>
			{isLoading && !recommendedUsers ? (
				<span>loading 6 Users specially for your Taste</span>
			) : (
				<Grid variant="row" gap="S" marginBelow="M">
					<div className="mb-8">
						<div className="flex flex-row flex-wrap -m-2">
							{pureRecommendedUsers.map((user) => (
								<div key={user.id} className="flex-initial w-4/12 p-3">
									<UserCard key={user.id} user={user} />
								</div>
							))}
						</div>
					</div>
				</Grid>
			)}
		</>
	);
};
