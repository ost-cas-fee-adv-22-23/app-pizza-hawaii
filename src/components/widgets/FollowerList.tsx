import { useSession } from 'next-auth/react';
import React, { useEffect, useState } from 'react';

import { useFolloweeContext } from '../../context/useFollowee';
import { services } from '../../services';
import { TUser } from '../../types';
import { UserCardList } from '../user/UserCardList';

export const FollowerList = () => {
	const { followees } = useFolloweeContext();
	const [users, setUsers] = useState<TUser[]>([]);

	const { data: session } = useSession();
	const accessToken = session?.accessToken;

	useEffect(() => {
		const fetchUsers = async () => {
			if (accessToken) {
				try {
					const users = await services.api.users.users({
						userIds: followees,
					});
					setUsers(users);
				} catch (error) {
					console.error(error);
				}
			}
			return () => {
				setUsers([]);
			};
		};

		fetchUsers();
	}, [accessToken, followees]);

	return <UserCardList users={users} />;
};
