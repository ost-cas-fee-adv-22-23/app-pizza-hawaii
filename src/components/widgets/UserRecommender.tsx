import { Headline } from '@smartive-education/pizza-hawaii';
import { useSession } from 'next-auth/react';
import React, { FC, useEffect, useState } from 'react';

import { services } from '../../services';
import { TUser } from '../../types';
import { UserCardList } from '../user/UserCardList';

/*
 * This component is used to display a list of users that the current user might want to follow or already follows.
 * We show a loading skeleton while the data is being fetched. We dont show the own user in the list.
 */

type TUserRecommender = {
	currentUserId: string;
	excludeUserIds?: string[];
	limit?: number;
};

export const UserRecommender: FC<TUserRecommender> = ({ currentUserId, excludeUserIds, limit = 6 }: TUserRecommender) => {
	const [isLoading, setIsLoading] = useState(true);
	const [recommendedUsers, setRecommendedUsers] = useState<TUser[]>([]);

	const { data: session } = useSession();
	const accessToken = session?.accessToken;

	useEffect(() => {
		(async () => {
			if (accessToken && currentUserId && limit) {
				try {
					const users = await services.api.users.recommendations({
						currentUserId,
						excludeUserIds,
						limit,
					});
					setRecommendedUsers(users);
					setIsLoading(false);
				} catch (error) {
					console.error(error);
					setIsLoading(false);
				}
			}
		})();
	}, [accessToken, currentUserId, excludeUserIds, limit]);

	return (
		<>
			<Headline as="h2" level={2}>
				Empfohlene User
			</Headline>
			<UserCardList users={recommendedUsers} loadingItems={isLoading ? limit : 0} />
		</>
	);
};
