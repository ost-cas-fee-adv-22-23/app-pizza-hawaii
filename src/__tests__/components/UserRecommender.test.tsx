import '@testing-library/jest-dom';
import { cleanup, render } from '@testing-library/react';
import { useSession } from 'next-auth/react';

import { UserRecommender } from '../../components/widgets/UserRecommender';

/**
 * Unit tests for the UserRecommender component
 *
 * These tests are written to make sure that the UserRecommender component is rendered correctly.
 * Unfortunately we cannot render the UserCardList with the UserCard component, because the UserCard component uses the useSession hook.
 * and another hook in the FollowButton component. But we can test the UserRecommender component with the UserCardList component.
 *
 * We test the following on UserRecommender
 * 1. Render UserRecommender component with the correct title
 * 2. Render UserRecommender component with the correct title classes
 * 3. Render UserRecommender component with the correct number of UserCard Skeletons
 * 4. Render UserRecommender component with the correct number of UserCard Skeletons when limit is 4
 * 5. Match snapshot
 *
 **/

const props = {
	currentUserId: '1234567890',
	excludeUserIds: ['10000004', '10000005', '10000006'],
	limit: 6,
};

jest.mock('next-auth/react', () => ({
	useSession: jest.fn(),
}));

describe('UserRecommender component', () => {
	afterEach(cleanup);
	test('renders the title Header with the correct words and correct classes', async () => {
		const mockSession = {
			user: { name: 'Filiks Adamski', email: 'filiks.adamski@mumble.com', id: '1234567890' },
			expires: '1',
		};
		(useSession as jest.Mock).mockReturnValue([mockSession, false]);

		const { container } = render(<UserRecommender {...props} />);
		expect(container.querySelector('h2')).toHaveTextContent('Empfohlene User');
		expect(container.querySelector('h2')).toHaveClass('leading-tight text-3xl font-bold');
	});

	// Expecting 6 Skeletons to be rendered when the component is loading
	test('renders 6 UserCard components with skeletons', async () => {
		const { container } = render(<UserRecommender {...props} />);
		expect(container.getElementsByClassName('animate-pulse')).toHaveLength(6);
	});

	// if the limit is 4, only 4 skeletons should be rendered
	test('renders 4 UserCard components with skeletons when given limit is 4', async () => {
		// updating props for test: propsLimit4
		const propsLimit4 = {
			currentUserId: '1234567890',
			excludeUserIds: ['10000004', '10000005', '10000006'],
			limit: 4,
		};
		const { container } = render(<UserRecommender {...propsLimit4} />);
		expect(container.getElementsByClassName('animate-pulse')).toHaveLength(4);
	});

	test('renders to match snapshot', async () => {
		const { container } = render(<UserRecommender {...props} />);
		expect(container).toMatchSnapshot();
	});
});
