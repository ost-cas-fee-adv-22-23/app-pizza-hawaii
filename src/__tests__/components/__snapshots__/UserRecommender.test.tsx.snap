// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRecommender component renders to match snapshot 1`] = `
<div>
  <h2
    class="leading-tight overflow-hidden text-ellipsis text-3xl font-bold"
  >
    Empfohlene User
  </h2>
  <div
    class="flex w-full flex-row gap-x-4 gap-y-2 sm:gap-x-2 mb-16  "
  >
    <div
      class="flex-1 flex flex-row flex-wrap -m-2"
    >
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex-initial w-4/12 p-2 sm:w-1/2"
      >
        <div
          class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-s px-s"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex w-full flex-col gap-x-8 gap-y-4 sm:gap-x-4   items-center"
            >
              <div
                class="h-24 w-24 bg-gray-300 rounded-full"
              />
              <div
                class="flex flex-col gap-2"
              >
                <div
                  class="h-4 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
                <div
                  class="h-5 w-24 bg-gray-300 rounded-full max-w-[100px]"
                />
              </div>
              <div
                class="h-10 w-full bg-gray-400 rounded-lg max-w-[160px]"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
